bindings/src/main/java/io/pravega/storage/filesystem/FileSystemChunkStorage.java

```diff
@@ -0,0 +1,294 @@
+            long fileSize = channel.size();
+            if (fileSize != offset) {
+                throw new IndexOutOfBoundsException(String.format("fileSize (%d) did not match offset (%d) for chunk %s", fileSize, offset, handle.getChunkName()));
+            }
+
+            // Wrap the input data into a ReadableByteChannel, but do not close it. Doing so will result in closing
+            // the underlying InputStream, which is not desirable if it is to be reused.
+            ReadableByteChannel sourceChannel = Channels.newChannel(data);
+            while (length > 0) {
+                long bytesWritten = channel.transferFrom(sourceChannel, offset, length);
+                assert bytesWritten > 0 : "Unable to make any progress transferring data.";
+                offset += bytesWritten;
+                totalBytesWritten += bytesWritten;
+                length -= bytesWritten;
+            }
+            channel.force(true);
+        } catch (IOException e) {
+            throw convertExeption(handle.getChunkName(), "doWrite", e);
+        }
+        return (int) totalBytesWritten;
+    }
+
+    @Override
+    public int doConcat(ConcatArgument[] chunks) throws ChunkStorageException {
+        try {
+            int totalBytesConcated = 0;
+            Path targetPath = Paths.get(config.getRoot(), chunks[0].getName());
+            long offset = chunks[0].getLength();
+
+            for (int i = 1; i < chunks.length; i++) {
+                val source = chunks[i];
+                Preconditions.checkArgument(!chunks[0].getName().equals(source.getName()), "target and source can not be same.");
+                Path sourcePath = Paths.get(config.getRoot(), source.getName());
+                long length = chunks[i].getLength();
+                Preconditions.checkState(offset <= getFileSize(targetPath));
+                Preconditions.checkState(length <= getFileSize(sourcePath));
+                try (FileChannel targetChannel = getFileChannel(targetPath, StandardOpenOption.WRITE);
+                     RandomAccessFile sourceFile = new RandomAccessFile(String.valueOf(sourcePath), "r")) {
+                    while (length > 0) {
+                        long bytesTransferred = targetChannel.transferFrom(sourceFile.getChannel(), offset, length);
+                        offset += bytesTransferred;
+                        length -= bytesTransferred;
+                    }
+                    targetChannel.force(true);
+                    totalBytesConcated += length;
+                    offset += length;
+                }
+
+            }
+            return totalBytesConcated;
+        } catch (IOException e) {
+            throw convertExeption(chunks[0].getName(), "doConcat", e);
+        }
+    }
+
+    @Override
+    protected boolean doTruncate(ChunkHandle handle, long offset) throws UnsupportedOperationException {
+        throw new UnsupportedOperationException();
```

Comment:
Maybe this `throw new UnsupportedOperationException() ` should be the default implementation in the base class

Reply:
Ok. Makes sense.

Reply:
done